core/src/main/java/hudson/lifecycle/Lifecycle.java

```diff
@@ -136,6 +139,19 @@ public void verifyRestartable() throws RestartNotSupportedException {
         return INSTANCE;
     }
 
+    /**
+     * Sets the singleton instance.
+     * Appropriate for implementations defined in plugins,
+     * since {@link #get} may be called before plugins are initialized,
+     * and so it is not safe to pass a plugin-defined class to the system property {@code hudson.lifecycle}
+     * despite the use of {@link PluginManager#uberClassLoader}.
```

Comment:
While a veto-type method in `RestartListener` (as in https://github.com/jenkinsci/jenkins/pull/8555#discussion_r1344393857) would probably be best, it occurs to me that we do not really need a `set` method. It would suffice for https://github.com/jenkinsci/jenkins/blob/d1b847366fab775219af3418bf29e6dc0add928d/core/src/main/java/hudson/lifecycle/Lifecycle.java#L83-L86 to quietly fall back to the default, but then have an initializer run after plugins are loaded which tries again (and complains noisily if the class still cannot be found).

I do not think https://github.com/jenkinsci/jenkins/pull/8555#discussion_r1344320863 is a real issue; it just means that for example a plugin-provided lifecycle cannot override `onStop` for the corner case of a boot failure. (The CloudBees CI implementation only needs to override `restart`, which should never be called before startup is complete so far as I know.)